src/cmap/auth/mongodb_oidc/token_entry_cache.ts

```diff
@@ -0,0 +1,110 @@
+import type { OIDCMechanismServerStep1, OIDCRequestTokenResult } from '../mongodb_oidc';
+
+/* 5 minutes in milliseonds */
+const EXPIRATION_BUFFER_MS = 300000;
+/* Default expiration is now for when no expiration provided */
+const DEFAULT_EXPIRATION_SECS = 0;
+
+/** @internal */
+export class TokenEntry {
+  tokenResult: OIDCRequestTokenResult;
+  serverResult: OIDCMechanismServerStep1;
+  expiration: number;
+
+  /**
+   * Instantiate the entry.
+   */
+  constructor(
+    tokenResult: OIDCRequestTokenResult,
+    serverResult: OIDCMechanismServerStep1,
+    expiration: number
+  ) {
+    this.tokenResult = tokenResult;
+    this.serverResult = serverResult;
+    this.expiration = expiration;
+  }
+
+  /**
+   * The entry is still valid if the expiration is more than
+   * 5 minutes from the expiration time.
+   */
+  isValid() {
+    return this.expiration - Date.now() > EXPIRATION_BUFFER_MS;
+  }
+}
+
+/**
+ * Cache of OIDC token entries.
+ * @internal
+ */
+export class TokenEntryCache {
+  entries: Map<string, TokenEntry>;
+
+  constructor() {
+    this.entries = new Map();
+  }
+
+  /**
+   * Set an entry in the token cache.
+   */
+  addEntry(
+    address: string,
+    username: string,
+    tokenResult: OIDCRequestTokenResult,
+    serverResult: OIDCMechanismServerStep1
+  ): TokenEntry {
+    const entry = new TokenEntry(
+      tokenResult,
+      serverResult,
+      expirationTime(tokenResult.expiresInSeconds)
+    );
+    this.entries.set(cacheKey(address, username), entry);
+    return entry;
+  }
+
+  /**
+   * Clear the cache.
+   */
+  clear(): void {
+    this.entries.clear();
+  }
+
+  /**
+   * Delete an entry from the cache.
+   */
+  deleteEntry(address: string, username: string): void {
+    this.entries.delete(cacheKey(address, username));
+  }
+
+  /**
+   * Get an entry from the cache.
+   */
+  getEntry(address: string, username: string): TokenEntry | undefined {
+    return this.entries.get(cacheKey(address, username));
+  }
+
+  /**
+   * Delete all expired entries from the cache.
+   */
+  deleteExpiredEntries(): void {
+    for (const [key, entry] of this.entries) {
+      if (!entry.isValid()) {
+        this.entries.delete(key);
+      }
+    }
+  }
+}
+
+/**
+ * Get an expiration time in milliseconds past epoch. Defaults to immediate.
+ */
+function expirationTime(expiresInSeconds?: number): number {
+  return Date.now() + (expiresInSeconds ?? DEFAULT_EXPIRATION_SECS) * 1000;
+}
+
+/**
+ * Create a cache key from the address and username.
+ */
+function cacheKey(address: string, username: string): string {
+  return `${address}-${username}`;
+}
```

Comment:
Actually, just got reminded of this because of a conversation on our side of the OIDC auth, from [the spec](https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#caching-credentials):

>  The cache key MUST also include a hashes of the callback function(s), if relevant and possible in the driver language.

I think we’d want that here as well, right? (Where I assume `hash` in JS is just the function identity, could be implemented e.g. by `WeakMap<Function, number>` which just assigns incremental numbers to the callbacks and then includes that in the string returned from `cacheKey`)